README.md
# TestKafka

Minimal Kafka 0.8 runner suitable for integration testing.

Adapted from the excellent [poseidon](https://github.com/bpot/poseidon)'s integration tests.

## Installation

Add TestKafka to your application's Gemfile:

```ruby
gem 'test_kafka', '~> 0.1.1'
```

and bundle:

    $ bundle

## Usage

```ruby
require 'test_kafka'

cluster = TestKafka.start('/usr/local/kafka')
# or specify custom a temp directory and kafka/zk ports:
# cluster = TestKafka.start('/usr/local/kafka', '/tmp', 9092, 2181)

# ... interact with Kafka/ZK ...

cluster.with_interruption do
  # broker temporarily stopped in this block
end

# broker is back up

cluster.stop
```

## Requirements

* Kafka 0.8 or higher


CHANGELOG.md
# 0.1.1

* Fix launching of Kafka 0.8.1+ from homebrew

# 0.1.0

* Initial release


lib/test_kafka.rb
require 'fileutils'
require 'test_kafka/cluster'

module TestKafka
  def self.start(kafka_path, tmp_root='/tmp', kafka_port=9092, zk_port=2181)
    tmp_dir = tmp_root + "/test_kafka"
    FileUtils.rm_rf(tmp_dir)

    TestKafka::Cluster.new(kafka_path, tmp_dir, kafka_port, zk_port).start
  end
end


spec/spec_helper.rb
RSpec.configure do |config|
  config.treat_symbols_as_metadata_keys_with_true_values = true
  # config.run_all_when_everything_filtered = true
  # config.filter_run :focus
  config.order = 'random'
end

DEFAULT_KAFKA_PATH = "/usr/local/kafka"
KAFKA_PATH = ENV["KAFKA_PATH"] || DEFAULT_KAFKA_PATH

require 'test_kafka/java_runner'

jar_paths = TestKafka::JavaRunner::JAR_PATTERNS.flat_map { |pat|
  Dir.glob(KAFKA_PATH + "/" + pat)
}

if jar_paths.empty?
  fail "Could not find Kafka JARs. Set the environment variable KAFKA_PATH or install Kafka to /usr/local/kafka."
end

def running?(pid)
  Process.kill(0, pid)
  true
rescue Errno::ESRCH
  false
end

require 'poseidon'

def write_messages(port, messages)
  producer = Poseidon::Producer.new(["localhost:#{port}"],
                                    'test_producer')
  producer.send_messages(messages.map { |m|
    Poseidon::MessageToSend.new('topic1', m)
  })
end

def read_messages(port)
  consumer = Poseidon::PartitionConsumer.new(
    'test_consumer',
    'localhost',
    broker_port,
    'topic1',
    0,
    :earliest_offset
  )

  consumer.fetch.map(&:value)
end


spec/test_kafka_spec.rb
require 'spec_helper'
require 'test_kafka'
require 'fileutils'

describe TestKafka do
  describe '.start' do
    context 'with custom args' do
      let(:tmp_root) { '/tmp/test_kafka-test' }
      let(:broker_port) { 9093 }
      let(:zk_port) { 2182 }

      it 'initializes and starts a single-node test cluster' do
        cluster = TestKafka.start(KAFKA_PATH, tmp_root, broker_port, zk_port)

        messages = ['value1', 'value2']
        write_messages(broker_port, messages)

        read_messages(broker_port).should eql messages

        cluster.stop
      end

      it 'deletes an existing test_kafka temp directory if one exists' do
        tmp_dir = tmp_root + "/test_kafka"
        FileUtils.mkdir_p(tmp_dir)
        canary_path = tmp_dir + "/canary"
        File.write(canary_path, "chirp")

        cluster = TestKafka.start(KAFKA_PATH, tmp_root, broker_port, zk_port)

        File.exist?(canary_path).should be_false

        cluster.stop
      end
    end

    context 'with default args' do
      before(:all) { @cluster = TestKafka.start(KAFKA_PATH) }
      after(:all) { @cluster.stop }

      it 'uses /tmp as a temp root' do
        Dir.exist?('/tmp/test_kafka').should be_true
      end

      it 'uses kafka port 9092' do
        @cluster.broker.port.should eql 9092
      end

      it 'uses ZK port 2181' do
        @cluster.zookeeper.port.should eql 2181
      end
    end
  end
end


lib/test_kafka/java_runner.rb
require 'fileutils'
require 'socket' # daemon_controller needs this but currently doesn't require it
require 'daemon_controller'

module TestKafka
  class JavaRunner
    MAX_HEAP_SIZE = 512
    JAR_PATTERNS = [
      # Some Kafka builds from 0.8.0:
      "core/target/scala-*/*.jar",
      # 0.8.1+
      "libexec/core/build/**/*.jar"
    ]

    def initialize(id, tmp_dir, java_class, port, kafka_path, properties={})
      @id = id
      @tmp_dir = tmp_dir
      @java_class = java_class
      @port = port
      @kafka_path = kafka_path
      @properties = properties
    end

    attr_reader :tmp_dir, :java_class, :kafka_path

    def start
      write_properties
      run
    end

    def stop
      daemon_controller.stop
    end

    def with_interruption
      stop
      begin
        yield
      ensure
        start
      end
    end

    def pid
      data = File.read(pid_path)

      data.to_i if data
    end

    private

    def classpath
      JAR_PATTERNS.flat_map { |pat| Dir.glob(kafka_path + "/" + pat) }.join(":")
    end

    def java_command
      "exec java -Xmx#{MAX_HEAP_SIZE}M -server -cp #{classpath} #{java_class} #{config_path}"
    end

    def daemon_controller
      @dc ||= DaemonController.new(
        :identifier => @id,
        :start_command => "#{java_command} >>#{log_path} 2>&1 & echo $! > #{pid_path}",
        :ping_command => [:tcp, '127.0.0.1', @port],
        :pid_file => pid_path,
        :log_file => log_path,
        :start_timeout => 25
      )
    end

    def run
      FileUtils.mkdir_p(log_dir)
      FileUtils.mkdir_p(pid_dir)
      daemon_controller.start
    end

    def write_properties
      FileUtils.mkdir_p(config_dir)
      File.open(config_path, "w+") do |f|
        @properties.each do |k,v|
          f.puts "#{k}=#{v}"
        end
      end
    end

    def pid_path
      "#{pid_dir}/#{@id}.pid"
    end

    def pid_dir
      "#{tmp_dir}/pid"
    end

    def log_path
      "#{log_dir}/#{@id}.log"
    end

    def log_dir
      "#{tmp_dir}/log"
    end

    def config_path
      "#{config_dir}/#{@id}.properties"
    end

    def config_dir
      "#{tmp_dir}/config"
    end
  end
end


lib/test_kafka/broker_runner.rb
require 'test_kafka/java_runner'

module TestKafka
  class BrokerRunner
    DEFAULT_PROPERTIES = {
      "broker.id" => 0,
      "port" => 9092,
      "num.network.threads" => 2,
      "num.io.threads" => 2,
      "socket.send.buffer.bytes" => 1048576,
      "socket.receive.buffer.bytes" => 1048576,
      "socket.request.max.bytes" => 104857600,
      "log.dir" => "/tmp/kafka-logs",
      "num.partitions" => 1,
      "log.flush.interval.messages" => 10000,
      "log.flush.interval.ms" => 1000,
      "log.retention.hours" => 168,
      "log.segment.bytes" => 536870912,
      "log.cleanup.interval.mins" => 1,
      "zookeeper.connect" => "localhost:2181",
      "zookeeper.connection.timeout.ms" => 1000000,
      "kafka.metrics.polling.interval.secs" => 5,
      "kafka.metrics.reporters" => "kafka.metrics.KafkaCSVMetricsReporter",
      "kafka.csv.metrics.dir" => "/tmp/kafka_metrics",
      "kafka.csv.metrics.reporter.enabled" => "false",
    }

    def initialize(kafka_path, tmp_dir, id, port, zk_port, partition_count = 1)
      @id = id
      @port = port
      @jr = JavaRunner.new("broker_#{id}",
                           tmp_dir,
                           "kafka.Kafka",
                           port,
                           kafka_path,
                           DEFAULT_PROPERTIES.merge(
                             "broker.id" => id,
                             "port" => port,
                             "log.dir" => "#{tmp_dir}/kafka-logs_#{id}",
                             "kafka.csv.metrics.dir" => "#{tmp_dir}/kafka_metrics",
                             "num.partitions" => partition_count,
                             "zookeeper.connect" => "localhost:#{zk_port}"
                           ))
    end

    def pid
      @jr.pid
    end

    def start
      @jr.start
    end

    def stop
      @jr.stop
    end

    def without_process
      @jr.without_process { yield }
    end
  end
end


lib/test_kafka/broker.rb
require 'test_kafka/java_runner'

module TestKafka
  class Broker
    DEFAULT_PROPERTIES = {
      "broker.id" => 0,
      "port" => 9092,
      "num.network.threads" => 2,
      "num.io.threads" => 2,
      "socket.send.buffer.bytes" => 1048576,
      "socket.receive.buffer.bytes" => 1048576,
      "socket.request.max.bytes" => 104857600,
      "log.dir" => "/tmp/kafka-logs",
      "num.partitions" => 1,
      "log.flush.interval.messages" => 10000,
      "log.flush.interval.ms" => 1000,
      "log.retention.hours" => 168,
      "log.segment.bytes" => 536870912,
      "log.cleanup.interval.mins" => 1,
      "zookeeper.connect" => "localhost:2181",
      "zookeeper.connection.timeout.ms" => 1000000,
      "kafka.metrics.polling.interval.secs" => 5,
      "kafka.metrics.reporters" => "kafka.metrics.KafkaCSVMetricsReporter",
      "kafka.csv.metrics.dir" => "/tmp/kafka_metrics",
      "kafka.csv.metrics.reporter.enabled" => "false",
    }.freeze

    def initialize(kafka_path, tmp_dir, port, zk_port, broker_id=0, partition_count=1)
      @broker_id = broker_id
      @port = port
      @jr = JavaRunner.new("broker_#{broker_id}",
                           tmp_dir,
                           "kafka.Kafka",
                           port,
                           kafka_path,
                           DEFAULT_PROPERTIES.merge(
                             "broker.id" => broker_id,
                             "port" => port,
                             "log.dir" => "#{tmp_dir}/kafka-logs_#{broker_id}",
                             "kafka.csv.metrics.dir" => "#{tmp_dir}/kafka_metrics",
                             "num.partitions" => partition_count,
                             "zookeeper.connect" => "localhost:#{zk_port}"
                           ))
    end

    attr_reader :broker_id, :port

    def start
      @jr.start
    end

    def stop
      @jr.stop
    end

    def pid
      @jr.pid
    end

    def with_interruption(&block)
      @jr.with_interruption(&block)
    end
  end
end
